pymeasure/instruments/rohdeschwarz/fsl.py

```diff
@@ -0,0 +1,259 @@
+    def sweep_time(self, value):
+        if type(value) is str and value.upper() == "AUTO":
+            self.write("SWE:TIME:AUTO ON")
+        else:
+            self.write(f"SWE:TIME {value}")
+
+    continuous_sweep = Instrument.control(
+        "INIT:CONT?",
+        "INIT:CONT %s",
+        "Continuous (True) or single sweep (False)",
+        validator=strict_discrete_set,
+        values=[True, False],
+        set_process=lambda x: "ON" if x else "OFF",
+        get_process=lambda x: bool(x),
+    )
+
+    def single_sweep(self):
+        """Perform a single sweep with synchronization."""
+        self.write("INIT; *WAI")
+
+    def continue_single_sweep(self):
+        """Continue with single sweep with synchronization."""
+        self.write("INIT:CONM; *WAI")
+
+    # Traces ------------------------------------------------------------------
+
+    def read_trace(self, n_trace=1):
+        """
+        Read trace data.
+
+        :param n_trace: The trace number (1-6). Default is 1.
+        :return: Nnumpy arrays of the trace data (frequency and amplitude).
+        """
+        y = np.array(self.values("TRAC{n_trace}? TRACE{n_trace}"))
+        x = np.linspace(self.freq_start, self.freq_stop, len(y))
+        return x, y
+
+    trace_mode = Instrument.control(
+        "DISP:TRAC:MODE?",
+        "DISP:TRAC:MODE %s",
+        "Trace mode ('WRIT', 'MAXH', 'MINH', 'AVER' or 'VIEW'",
+        validator=strict_discrete_set,
+        values=["WRIT", "MAXH", "MINH", "AVER", "VIEW"],
+    )
+
+    # Markers -----------------------------------------------------------------
+
+    def create_marker(self, num=1, is_delta_marker=False):
+        """
+        Create a marker.
+
+        :param num: The marker number (1-4)
+        :param is_delta_marker: True if the marker is a delta marker, default
+            is False.
+        :return: The marker object.
+        """
+        return self.Marker(self, num, is_delta_marker)
+
+    class Marker:
+        def __init__(self, instrument, num, is_delta_marker):
+            """
+            Marker and Delte Marker class.
+
+            :param instrument: The FSL instrument.
+            :param num: The marker number (1-4)
+            :param is_delta_marker: True if the marker is a delta marker,
+                defaults to False.
+            """
+            self.instr = instrument
+            self.is_delta_marker = is_delta_marker
+            # Building the marker name for the commands.
+            if self.is_delta_marker:
+                # Smallest delta marker number is 2.
+                self.name = "DELT" + str(max(2, num))
+            else:
+                self.name = "MARK"
+                if num > 1:
+                    # Marker 1 doesn't get a number.
+                    self.name = self.name + str(num)
+
+            self.activate()
+
+        def activate(self):
+            """Activate a marker."""
+            self.instr.write(f"CALC:{self.name}:STAT ON")
```

Comment:
The way to avoid having to manually prefix this channel ID everywhere, and get a nicely composable channel/Marker with properties that you can create with the control/setting/measurement factories is to [implement read/write/ask methods that delegate to the instrument](https://github.com/pymeasure/pymeasure/blob/7580edd2cc5dfaafb9c0d6462d6f95cd87a3d468/pymeasure/instruments/tektronix/afg3152c.py#L127-L144). (although you're free to use modern f-strings instead of the % syntax, of course)

Reply:
Cool, I like this. Changed in 6ea22bb .